homeassistant/components/firmata/binary_sensor.py

```diff
@@ -0,0 +1,84 @@
+"""Support for Firmata binary sensor input."""
+
+import logging
+
+from homeassistant.components.binary_sensor import BinarySensorEntity
+from homeassistant.const import CONF_NAME
+
+from .const import CONF_NEGATE_STATE, CONF_PIN, CONF_PIN_MODE, DOMAIN
+from .pin import FirmataBinaryDigitalInput, FirmataPinUsedException
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Firmata binary sensors."""
+    new_entities = []
+
+    boards = hass.data[DOMAIN]
+    board = boards[config_entry.entry_id]
+    for binary_sensor in board.binary_sensors:
+        try:
+            binary_sensor_entity = FirmataBinarySensor(
+                board, config_entry, **binary_sensor
+            )
+            new_entities.append(binary_sensor_entity)
+        except FirmataPinUsedException:
+            _LOGGER.error(
+                "Could not setup binary sensor on pin %s since pin already in use.",
+                binary_sensor[CONF_PIN],
+            )
+    async_add_entities(new_entities)
+
+
+class FirmataBinarySensor(BinarySensorEntity):
+    """Representation of a binary sensor on a Firmata board."""
+
+    def __init__(self, board, config_entry, **kwargs):
+        """Initialize the binary sensor."""
+        self._name = kwargs[CONF_NAME]
+        self._state = None
+        self._config_entry = config_entry
+        self._conf = kwargs
+
+        pin = self._conf[CONF_PIN]
+        pin_mode = self._conf[CONF_PIN_MODE]
+        negate = self._conf[CONF_NEGATE_STATE]
+
+        self._location = (DOMAIN, self._config_entry.entry_id, "pin", pin)
+        self._unique_id = "_".join(str(i) for i in self._location)
+
+        self._api = FirmataBinaryDigitalInput(board, pin, pin_mode, negate)
```

Comment:
Please create the api first outside the entity and pass the api instance to the entity as an argument on instantiation.

Reply:
So I should iterate through the config and setup these classes in the integration's `async_setup_entry` before I call `hass.config_entries.async_forward_entry_setup(config_entry, "binary_sensor")`?

Reply:
No we can create the pin api in the platform `async_setup_entry`.

Reply:
Yes, I think that is what I meant.

Reply:
The platforms are the binary sensor and switch modules.